src/customFileParser/main.ts

```diff
@@ -1,81 +1,61 @@
-  sectors[sectorMetadata.sectorId] = rootSector;
+  const idToSectorMap: {[name: string]: Sector} = {};
+  const sceneStats: SceneStats = {
+    numInstancedMeshes: 0,
+    numMergedMeshes: 0,
+  };
+  const mergedMeshMap: any = {};
+  const treeIndexNodeIdMap: number[] = [];
+  const colorMap: THREE.Color[] = [];
+
+  let rootSector = undefined;
+  let uncompressedValues = undefined;
+  const sectorPathToPrimitiveData: any = {};
+  const sectorPathToInstancedMeshData: any = {};
+  const sectorPathToMergedMeshData: any = {};
 
   while (fileReader.location < fileBuffer.byteLength) {
-    sectorStartLocation = fileReader.location;
-    sectorByteLength = fileReader.readUint32();
-    sectorMetadata = fileReader.readSectorMetadata(sectorByteLength);
-    geometryIndexHandlers = fileReader.readSectorGeometryIndexHandlers(sectorStartLocation + sectorByteLength);
-    const newSectorObject = createSector(sectorMetadata, uncompressedValues, geometryIndexHandlers);
-    const parentSector = sectors[sectorMetadata.parentSectorId];
-    if (parentSector !== undefined) {
-      parentSector.addChild(newSectorObject);
-      parentSector.object3d.add(newSectorObject.object3d);
-      sectors[sectorMetadata.sectorId] = newSectorObject;
-    } else { throw Error('Parent sector not found'); }
+    const sectorStartLocation = fileReader.location;
+    const sectorByteLength = fileReader.readUint32();
+    const sectorMetadata = fileReader.readSectorMetadata(sectorByteLength);
+    const sector = new Sector(sectorMetadata.sectorBBoxMin, sectorMetadata.sectorBBoxMax);
+    idToSectorMap[sectorMetadata.sectorId] = sector;
+
+    if (rootSector === undefined || uncompressedValues === undefined) {
+      rootSector = sector;
+      uncompressedValues = fileReader.readUncompressedValues();
+    } else {
+      const parentSector = idToSectorMap[sectorMetadata.parentSectorId];
+      if (parentSector !== undefined) {
+        parentSector.addChild(sector);
+        parentSector.object3d.add(sector.object3d);
+      } else { throw Error('Parent sector not found'); }
+    }
+    const compressedGeometryData = fileReader.readCompressedGeometryData(sectorStartLocation + sectorByteLength);
+    sectorPathToPrimitiveData[sector.path] = compressedGeometryData.primitives;
+    sectorPathToInstancedMeshData[sector.path] = compressedGeometryData.instancedMesh;
+    sectorPathToMergedMeshData[sector.path] = compressedGeometryData.mergedMesh;
   }
 
-  return rootSector;
-}
+  unpackPrimitives(rootSector!, uncompressedValues!, sectorPathToPrimitiveData,
+    treeIndexNodeIdMap, colorMap);
+  unpackInstancedMeshes(rootSector!, uncompressedValues!, sectorPathToInstancedMeshData, sceneStats,
+    treeIndexNodeIdMap, colorMap);
+  unpackMergedMeshes(rootSector!, uncompressedValues!, sectorPathToMergedMeshData, sceneStats,
+    treeIndexNodeIdMap, colorMap);
+
+  const sectors = idToSectorMap;
+  const nodeIdTreeIndexMap: {[s: number]: number} = {};
```

Comment:
We have a type `TreeIndexNodeIdMap` in `parser/parseUtils`. Should we create one for NodeIdTreeIndexMap there as well? Or is this doing typescript too much? (This is eventually being exposed isn't it though?)